### PR TITLE
Core functionality

### PR DESCRIPTION
**Pull request by LShostenko**

### CODE REVIEW

File: keywords_similarity/__init__.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from keywords_similarity.keywords_similarity import (
+    keywords2synsets, keywords_semantic_similarity, keywords_string_similarity,
+)
+
+__all__ = [
+    'keywords2synsets',
```
Comment (pcinkh) [inline keywords_similarity/__init__.py]: I think it is not so good idea to use "2"

File: keywords_similarity/keywords_similarity.py [hunk 2]
```diff
@@ -1,14 +1,76 @@
...
+    for i, j in product(range(n_1), range(n_2)):
+        similarity_matrix[i, j] = similarity_function(group_1[i], group_2[j])
+
+    np.nan_to_num(similarity_matrix, copy=False)
 
     return similarity_matrix
+
+
+def keywords_semantic_similarity(
+    keywords_1,
```
Comment (pcinkh) [inline keywords_similarity/keywords_similarity.py]: think about left/right

File: keywords_similarity/keywords_similarity.py [hunk 3]
```diff
@@ -1,14 +1,76 @@
...
+
+    for keywords in (keywords_1, keywords_2):
+        synset = keywords2synsets(
+            keywords,
+            only_nouns=only_nouns,
+            keep_duplicates=keep_duplicates,
+        )
+
+        if not synset:
+            warnings.warn(
```
Comment (pcinkh) [inline keywords_similarity/keywords_similarity.py]: need to discuss

File: keywords_similarity/keywords_similarity.py [hunk 4]
```diff
@@ -1,14 +1,76 @@
...
+
+
+def keywords_string_similarity(
+    keywords_1,
+    keywords_2,
+    similarity_function=Levenshtein().get_sim_score,
+    greedy=False,
+):
+    if not keywords_1 or not keywords_2:
+        warnings.warn('empty keywords', RuntimeWarning)
```
Comment (pcinkh) [inline keywords_similarity/keywords_similarity.py]: the same

File: keywords_similarity/keywords_similarity.py (hunk 3 above)
Comment (pcinkh) [reply keywords_similarity/keywords_similarity.py]: silent as a solution

File: keywords_similarity/keywords_similarity.py [hunk 5]
```diff
@@ -1,14 +1,76 @@
...
+
+    np.nan_to_num(similarity_matrix, copy=False)
 
     return similarity_matrix
+
+
+def keywords_semantic_similarity(
+    keywords_1,
+    keywords_2,
+    similarity_metric='wup',
```
Comment (pcinkh) [inline keywords_similarity/keywords_similarity.py]: check for valid option

File: keywords_similarity/wn.py [hunk 6]
```diff
@@ -1,29 +1,31 @@
...
+
 from nltk.corpus import wordnet as wn
 
-SIMILARITY_METHODS = ['jcn', 'lin', 'path', 'wup']
+SIMILARITY_METHODS = ('path', 'path_cached', 'wup', 'wup_cached')
 
 
-def get_similarity_method(method):
+def get_similarity_function(method):
     method = method.lower()
```
Comment (pcinkh) [inline keywords_similarity/wn.py]: redundant transformation

File: keywords_similarity/wn.py [hunk 7]
```diff
@@ -36,4 +38,21 @@ def keywords2synsets(keywords, only_nouns=True):
         if word_synonyms:
             synsets.append(word_synonyms[0])
 
+    if not keep_duplicates:
+        synsets = sorted(list(set(synsets)), key=synsets.index)
+
     return synsets
+
+
+@lru_cache(maxsize=1048576)
```
Comment (pcinkh) [inline keywords_similarity/wn.py]: oO

File: keywords_similarity/wn.py [hunk 8]
```diff
@@ -36,4 +38,21 @@ def keywords2synsets(keywords, only_nouns=True):
...
+def path_similarity_cached(synset_1, synset_2):
+    args = sorted([synset_1, synset_2])
```
Comment (pcinkh) [inline keywords_similarity/wn.py]: synsets

**Comment by LShostenko** [comment]
Major changes, need a review again

**Pull request merged**